driver-core/src/main/com/mongodb/client/model/expressions/ArrayExpression.java

```diff
@@ -16,6 +16,12 @@
 
 package com.mongodb.client.model.expressions;
 
+/**
+ * Expresses an array value. An array value is a finite, ordered collection of
+ * elements of a certain type.
+ *
+ * @param <T> the type of the elements in the array
+ */
```

Comment:
Since expressions are evaluated on the server, and the server can operate only with BSON types, it does not seem possible to abstract expression types of BSON types: if you do that, then a user can

- neither express a value stored in MongoDB via the expression API as it's unspecified what type should be used
- nor store an expression value to MongoDB as it is unclear how it will be represented there (for example, with the current lack of specification, you may be expressing all values as BSON `String` values that contain extended JSON).

So it appears to me that if an expression type has a BSON counterpart (which is true for everything but the `Map` type you proposed), it should be documented as such.

Reply:
Discussed in Zoom, decided that `MqlExpressionCodecProvider` is the right place to specify the relation between the abstract types whose values can be represented by the expression API and the types of a specific execution context, e.g., MongoDB. Fortunately, we also [decided](https://github.com/mongodb/mongo-java-driver/pull/1025#discussion_r1009936352) that `MqlExpressionCodecProvider` has to be public API.